lonestar/analytics/cpu/random-walk/RandomWalk.cpp

```diff
@@ -0,0 +1,631 @@
+          transformed_num_edge_types_walks =
+              TransformVectors(num_edge_types_walks);
+
+      katana::gstl::Vector<double> means =
+          ComputeMeans(transformed_num_edge_types_walks);
+
+      ComputeTransitionMatrix(transformed_num_edge_types_walks, means);
+    }
+  }
+};
+
+template <typename Graph>
+void
+InitializeDegrees(const Graph& graph, katana::LargeArray<uint64_t>* degree) {
+  katana::do_all(
+      katana::iterate(graph),
+      [&](GNode n) {
+        (*degree)[n] = std::distance(graph.edge_begin(n), graph.edge_end(n));
+      },
+      katana::steal());
+}
+
+void
+PrintWalks(
+    const katana::InsertBag<katana::gstl::Vector<uint32_t>>& walks,
+    const std::string& output_file) {
+  std::ofstream f(output_file);
+
+  for (auto walk : walks) {
+    for (auto node : walk) {
+      f << node << " ";
+    }
+    f << std::endl;
+  }
+}
+
+template <typename Algo>
+void
+run() {
+  using Graph = typename Algo::Graph;
+
+  Algo algo;
+
+  katana::gInfo("Reading from file: ", inputFile, "\n");
+  std::unique_ptr<katana::PropertyFileGraph> pfg =
+      MakeFileGraph(inputFile, edge_property_name);
+
+  auto pg_result = katana::PropertyGraph<
+      typename Algo::NodeData, typename Algo::EdgeData>::Make(pfg.get());
+
+  if (!pg_result) {
+    KATANA_LOG_FATAL("could not make property graph: {}", pg_result.error());
+  }
+
+  auto res = katana::SortAllEdgesByDest(pfg.get());
+  if (!res) {
+    KATANA_LOG_FATAL("Sorting property file graph failed: {}", res.error());
+  }
+
+  Graph graph = pg_result.value();
+
+  katana::gInfo(
+      "Read ", graph.num_nodes(), " nodes, ", graph.num_edges(), " edges\n");
+
+  katana::gPrint("size: ", graph.size(), "\n");
+
+  katana::InsertBag<katana::gstl::Vector<uint32_t>> walks;
+
+  katana::gInfo("Starting random walks...");
+  katana::StatTimer execTime("Timer_0");
+  execTime.start();
+
+  katana::LargeArray<uint64_t> degree;
+  degree.allocateBlocked(graph.size());
+  InitializeDegrees<Graph>(graph, &degree);
+
+  algo.RandomWalks(graph, &walks, degree);
+
+  degree.destroy();
+  degree.deallocate();
+
+  execTime.stop();
+
+  if (output) {
+    std::string output_file = outputLocation + "/" + outputFile;
+    katana::gInfo("Writing random walks to a file: ", output_file);
+    PrintWalks(walks, output_file);
+  }
+}
+
+int
+main(int argc, char** argv) {
+  std::unique_ptr<katana::SharedMemSys> G =
+      LonestarStart(argc, argv, name, desc, nullptr, &inputFile);
+
+  katana::StatTimer totalTime("TimerTotal");
+  totalTime.start();
+
+  if (!symmetricGraph) {
+    KATANA_DIE(
+        "This application requires a symmetric graph input;"
+        " please use the -symmetricGraph flag "
+        " to indicate the input is a symmetric graph.");
+  }
+
+  katana::gInfo("Only considering unweighted graph currently");
+
+  switch (algo) {
+  case Algo::node2vec:
+    run<Node2VecAlgo>();
+    break;
+  case Algo::edge2vec:
+    run<Edge2VecAlgo>();
```

Comment:
(CHange for later) FYI: Edge2vec expects edge_type (uint32_t) property. I think we will have to write a pass to find the distinct number of edge types instead of expecting it from the users. I added commandline argument for simplicity and we can change that later.